### PR TITLE
kubeadm add phase runner

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR implements the Phase runner as defined in [KEP0029](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0029-20180918-kubeadm-phases-beta.md) 

**Which issue(s) this PR fixes** : 
Rif https://github.com/kubernetes/kubeadm/issues/1163

**Special notes for your reviewer**:

/sig cluster-lifecycle
/kind documentation
/assign @neolit123 
/cc @timothysc 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@chuckha 
@rosti 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by neolit123** [comment]
thanks, best to rename the title as there are now 2 PRs `kubeadm add phase runner`

**Review by timothysc — CHANGES_REQUESTED** [review]
So I'm a little concerned we're over-engineering for what we need at this point, and could leverage existing libraries to some of the internal logistics.

File: cmd/kubeadm/app/cmd/phases/workflow/phase.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/*
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/phase.go]: In order to simplify I removed the interface and now there is only a struct

**Comment by fabriziopandini** [comment]
@timothysc I further simplified things removing the interface; let me know if now it is better

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+		return nil
+	})
+	line += "```"
+	return line
+}
+
+// BindToCommand bind the Runner to a cobra command by altering
+// command help, adding phase related flags and by adding phases subcommands
+func (e *Runner) BindToCommand(cmd *cobra.Command) {
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/workflow/runner.go]: Please add a comment that order matters here and that this needs to be done at the end once your phases are all added.

**Review by timothysc — CHANGES_REQUESTED** [review]
/approve 

minor comments then lgtm and lets keep moving.

File: cmd/kubeadm/app/cmd/phases/workflow/phase.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	Short string
+
+	// Long returns the long description of the phase.
+	Long string
+
+	// Example returns the example for the phase.
+	Example string
+
+	// Hidden define if the phase should be hidden in the workflow help.
+	Hidden bool
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/phase.go]: done

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/runner.go]: done

**Comment by neolit123** [comment]
LGTM, thanks @fabriziopandini

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by timothysc** [comment]
@fabriziopandini - I look forward to the refactor!

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+type RunnerOptions struct {
+	// FilterPhases defines the list of phases to be executed (if empty, all).
+	FilterPhases []string
+
+	// SkipPhases defines the list of phases to be excluded by execution (if empty, none).
+	SkipPhases []string
+}
+
+// RunData defines the data shared among all the phases included in the workflow, that is any type.
+type RunData = interface{}
```
Comment (chuckha) [inline cmd/kubeadm/app/cmd/phases/workflow/runner.go]: What is the benefit of the type alias here?
 
versus type definition like this 

```
type RunData interface{}
```

Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/runner.go]: @chuckha 
I simply found ugly using interface{}, and thus created an alias for it.

**Pull request merged**